File: src/lib/cc/python/ISC/CC/session.py
# Copyright (C) 2009  Internet Systems Consortium.
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND INTERNET SYSTEMS CONSORTIUM
# DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL
# INTERNET SYSTEMS CONSORTIUM BE LIABLE FOR ANY SPECIAL, DIRECT,
# INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING
# FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
# NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION
# WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

import sys
import socket
import struct

from ISC.CC import Message

class ProtocolError(Exception): pass
class NetworkError(Exception): pass
class SessionError(Exception): pass

class Session:
    def __init__(self, port=9912):
        self._socket = None
        self._lname = None
        self._recvbuffer = bytearray()
        self._recvlength = None
        self._sendbuffer = bytearray()
        self._sequence = 1

        try:
            self._socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self._socket.connect(tuple(['127.0.0.1', port]))

            self.sendmsg({ "type": "getlname" })
            env, msg = self.recvmsg(False)
            self._lname = msg["lname"]
            if not self._lname:
                raise ProtocolError("Could not get local name")
        except socket.error as se:
                raise SessionError(se)

    @property
    def lname(self):
        return self._lname

    def sendmsg(self, env, msg = None):
        if type(env) == dict:
            env = Message.to_wire(env)
        if type(msg) == dict:
            msg = Message.to_wire(msg)
        self._socket.setblocking(1)
        length = 2 + len(env);
        if msg:
            length += len(msg)
        self._socket.send(struct.pack("!I", length))
        self._socket.send(struct.pack("!H", len(env)))
        self._socket.send(env)
        if msg:
            self._socket.send(msg)

    def recvmsg(self, nonblock = True):
        data = self._receive_full_buffer(nonblock)
        if data and len(data) > 2:
            header_length = struct.unpack('>H', data[0:2])[0]
            data_length = len(data) - 2 - header_length
            if data_length > 0:
                return Message.from_wire(data[2:header_length+2]), Message.from_wire(data[header_length + 2:])
            else:
                return Message.from_wire(data[2:header_length+2]), None
        return None, None

    def _receive_full_buffer(self, nonblock):
        if nonblock:
            self._socket.setblocking(0)
        else:
            self._socket.setblocking(1)

        if self._recvlength == None:
            length = 4
            length -= len(self._recvbuffer)
            try:
                data = self._socket.recv(length)
            except:
                return None
            if data == "": # server closed connection
                raise ProtocolError("Read of 0 bytes: connection closed")

            self._recvbuffer += data
            if len(self._recvbuffer) < 4:
                return None
            self._recvlength = struct.unpack('>I', self._recvbuffer)[0]
            self._recvbuffer = bytearray()

        length = self._recvlength - len(self._recvbuffer)
        while (length > 0):
            try:
                data = self._socket.recv(length)
            except:
                return None
            if data == "": # server closed connection
                raise ProtocolError("Read of 0 bytes: connection closed")
            self._recvbuffer += data
            length -= len(data)
        data = self._recvbuffer
        self._recvbuffer = bytearray()
        self._recvlength = None
        return (data)

    def _next_sequence(self):
        self._sequence += 1
        return self._sequence

    def group_subscribe(self, group, instance = "*", subtype = "normal"):
        """subtype can be 'normal' or 'meonly'"""
        self.sendmsg({
            "type": "subscribe",
            "group": group,
            "instance": instance,
            "subtype": subtype,
        })

    def group_unsubscribe(self, group, instance = "*"):
        self.sendmsg({
            "type": "unsubscribe",
            "group": group,
            "instance": instance,
        })

    def group_sendmsg(self, msg, group, instance = "*", to = "*"):
        seq = self._next_sequence()
        self.sendmsg({
            "type": "send",
            "from": self._lname,
            "to": to,
            "group": group,
            "instance": instance,
            "seq": seq,
        }, Message.to_wire(msg))
        return seq

    def group_recvmsg(self, nonblock = True):
        env, msg  = self.recvmsg(nonblock)
        if env == None:
            # return none twice to match normal return value
            # (so caller won't get a type error on no data)
            return (None, None)
        return (msg, env)

    def group_reply(self, routing, msg):
        seq = self._next_sequence()
        self.sendmsg({
            "type": "send",
            "from": self._lname,
            "to": routing["from"],
            "group": routing["group"],
            "instance": routing["instance"],
            "seq": seq,
            "reply": routing["seq"],
        }, Message.to_wire(msg))
        return seq

if __name__ == "__main__":
    import doctest
    doctest.testmod()

File: src/lib/cc/ruby/lib/cc/session.rb
# Copyright (C) 2009  Internet Systems Consortium.
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND INTERNET SYSTEMS CONSORTIUM
# DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL
# INTERNET SYSTEMS CONSORTIUM BE LIABLE FOR ANY SPECIAL, DIRECT,
# INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING
# FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
# NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION
# WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

require 'socket'

class CC
class ProtocolError < Exception ; end
end

class CC
class Session
  attr_reader :socket
  attr_reader :lname

  #
  # :host => host to connect to (defaults to "127.0.0.1")
  # :port => port to connect to (defaults to 9913)
  #
  def initialize(args = {})
    @socket = nil      # TCP socket.
    @lname = nil       # local name, or nil if not connected.
    @recvbuffer = ""   # data buffer for partial reads.
    @recvlength = nil  # if non-nil, we have a length to fill buffer to.
    @sendbuffer = ""   # pending output data.
    @sequence = "a"    # per message sequence id, always unique

    options = {
      :host => "127.0.0.1",
      :port => 9912
    }.merge(args)

    @socket = TCPSocket.new(options[:host], options[:port])

    #
    # Get our local name.
    #
    sendmsg({ :type => :getlname })
    msg = recvmsg(false)
    @lname = msg["lname"]
    if @lname.nil?
      raise CC::ProtocolError, "Could not get local name"
    end
    CC::set_utf8(@lname)
  end

  #
  # Send a message to the controller.  The item to send can either be a
  # CC::Message object, or a Hash.  If a Hash, it will be internally
  # converted to a CC::Message before transmitting.
  #
  # A return value of true means the entire message was not
  # transmitted, and a call to send_pending will have to be
  # made to send remaining data.  This should only happen when
  # the socket is in non-blocking mode.
  #
  def sendmsg(msg)
    if msg.is_a?(Hash)
      msg = CC::Message.new(msg)
    end

    unless msg.is_a?(CC::Message)
      raise ArgumentError, "msg is not a CC::Message or a Hash"
    end

    wire = msg.to_wire
    @sendbuffer << [wire.length].pack("N")
    @sendbuffer << wire

    send_pending
  end

  #
  # Send as much data as we can.  
  def send_pending
    return false if @sendbuffer.length == 0
    sent = @socket.send(@sendbuffer, 0)
    @sendbuffer = @sendbuffer[sent .. -1]
    @sendbuffer.length == 0 ? true : false
  end

  def recvmsg(nonblock = true)
    data = receive_full_buffer(nonblock)
    if data
      CC::Message::from_wire(data)
    else
      nil
    end
  end

  def group_subscribe(group, instance = "*", subtype = "normal")
    sendmsg({ :type => "subscribe",
              :group => group,
              :instance => instance,
              :subtype => subtype,
            })
  end

  def group_unsubscribe(group, instance = "*")
    sendmsg({ :type => "unsubscribe",
              :group => group,
              :instance => instance,
            })
  end

  def group_sendmsg(msg, group, instance = "*", to = "*")
    seq = next_sequence
    sendmsg({ :type => "send",
              :from => @lname,
              :to => to,
              :group => group,
              :instance => instance,
              :seq => seq,
              :msg => CC::Message.to_wire(msg),
            })
    seq
  end

  def group_replymsg(routing, msg)
    seq = next_sequence
    sendmsg({ :type => "send",
              :from => @lname,
              :to => routing["from"],
              :group => routing["group"],
              :instance => routing["instance"],
              :seq => seq,
              :reply => routing["seq"],
              :msg => CC::Message.to_wire(msg),
            })
    seq
  end

  def group_recvmsg(nonblock = true)
    msg = recvmsg(nonblock)
    return nil if msg.nil?
    data = CC::Message::from_wire(msg["msg"])
    msg.delete("msg")
    return [data, msg]
  end

  private

  def next_sequence
    @sequence.next!
  end

  #
  # A rather tricky function.  If we have something waiting in our buffer,
  # and it will satisfy the current request, we will read it all in.  If
  # not, we will read only what we need to satisfy a single message.
  #
  def receive_full_buffer(nonblock)
    # read the length prefix if we need it still.
    if @recvlength.nil?
      length = 4
      length -= @recvbuffer.length
      data = nil
      begin
        if nonblock
          data = @socket.recv_nonblock(length)
        else
          data = @socket.recv(length)
        end
        rescue Errno::EINPROGRESS
        rescue Errno::EAGAIN
      end
      return nil if data == nil
      @recvbuffer += data
      return nil if @recvbuffer.length < 4
      @recvlength = @recvbuffer.unpack('N')[0]
      @recvbuffer = ""
      CC::set_binary(@recvbuffer)
    end

    #
    # we have a length target.  Loop reading data until we get enough to
    # fill our buffer.
    #
    length = @recvlength - @recvbuffer.length
    while (length > 0) do
      data = nil
      begin
        if nonblock
          data = @socket.recv_nonblock(length)
        else
          data = @socket.recv(length)
        end
        rescue Errno::EINPROGRESS
        rescue Errno::EAGAIN
      end
      return nil if data == 0 # blocking I/O
      @recvbuffer += data
      length -= data.length
    end

    data = @recvbuffer
    @recvbuffer = ""
    @recvlength = nil
    data
  end

end # class Session
end # class CC
